lib/spack/spack/solver/asp.py

```diff
@@ -468,8 +468,14 @@ def _compute_specs_from_answer_set(self):
 
             if candidate and candidate.satisfies(input_spec):
                 self._concrete_specs.append(answer[node])
-                self._concrete_specs_by_input[input_spec] = answer[node]
+                self._concrete_specs_by_input.append((input_spec, answer[node]))
```

Comment:
Notes:

* The only things that use `concrete_specs_by_input` call `update` with it, so it doesn't have to be a dictionary
* Without this, `test_concretize_partial_old_dag_hash_spec` fails (see https://github.com/spack/spack/actions/runs/7894899814/job/21546544138?pr=42655)
    * Overall, we shouldn't hash specs before calling `_finalize_concretization`
    * I don't think it makes sense to call `finalize_concretization` until we know the solver has succeeded
    * So to resolve this cycle, I removed the need for hashing by making this a `List[Tuple[Spec, Spec]]`, I considered adding an explicit typehint, but I'm concerned that would create a cyclic import (Specs need Solvers need Specs)

Reply:
I'm revisiting this and undoing it because input/concrete specs should always be hashable. The test failure was more an issue with how the test was simulating a dependency without a package hash, so I have addressed that directly.